content/blog/2019/06/2019-06-06-cloudfeature-gsoc-cp1.adoc

```diff
@@ -0,0 +1,104 @@
+opengraph:
+  image: /images/folder/icon.png # optional
+---
+
+=== About Me
+My name is Yufei Zhang, a student from Zhejiang University and Simon Fraser University Duel Degree Program. I want to share the discoveries when we are exploring the possiblities of extending current external workspace manager plugin (referred to as EWM) to support AWS EFS. 
+
+=== Backgrounds
+For users, they might want to provision a high-available Jenkins cluster, together with other services such as Keycloak for SSO, and use k8s to autoscale agents. All these operations needs a lot of configuration and cloud resources if they decide to base Jenkins on clouds. 
+
+In such cases, EWM was designed to give multiple agents access to shared disks, hence reduce the cost of copying duplicated data, such as codes and outputs in the workspace, or archieves that needs to be managed by other services(e.g. Some companies might want use archieves to do some analysising). With these backgrounds, it is natural to think of "supporting cloud disks". 
+
+Our project proposal was based on extending cloud features, but then we discovered we were on the wrong track maybe.
+
+=== Discoveries
+In order to make things clear, we need to understand the true nature of EWM. As we stated earlier, we want multiple agents to share disk spaces, and the main obstacle was they all have their own workspace directory(paths) by default. If we can modify the default workspace directory, then we might be able to utilize the shared NFS(Network File System).  And that is what EWM did. There are other features in like different strategies choosing a disk, but the way EWM to solve the problem is simple but effective.
+
+And by the way, from my perspective,the main problem, synchronization when visiting shared resources, is guaranteed by the implementation of the NFS itself, and the EWM didn't offer any synchronization mechanism to support concurrent execution. Jenkins itself is not designed to be synchronized. It is left to the users to partition their jobs and let the user manage syncronization. There might be some tricks to do that, but not within the scope of this blog.
+
+When we know what EWM did, we also realized : The Cloud feature is actually already supported. First, from the users perspective, they don't need to worry about the type of the disk, no matter it is a ext2 or ext3 or NFS or any other file system, they all look the same to host machine because of the concept of Virtual File System. If user can mount a disk to their host machine, then the EWM would be able to use them. So user can just create an EFS(Elastic File System, from AWS) in using whatever tools they like, and mount it. Then they can orchestra EFS with EWM.
+
+Then the problem becomes, should Jenkins be responsible to manage the NFS, for example, creating/managing/mounting/destroying the AWS EFS ? Well, my answer is not. There are some reasons. We will use AWS as example, but the same for other cloud providers.
```

Comment:
"creating/managing/mounting/destroying" -> "creating, managing, mounting and destroying"